crates/ghost_actor/src/ghost_protocol.rs

```diff
@@ -0,0 +1,169 @@
+pub enum GhostProtocolDestination {
+    Actor,
+    Owner,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+/// The type of message. Responses are directed toward the original requestor
+pub enum GhostProtocolVariantType {
+    Event,
+    Request,
+    Response,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+/// Identifies the *type* of variant, without considering the content
+pub struct GhostProtocolDiscriminant {
+    pub id: &'static str,
+    pub destination: GhostProtocolDestination,
+    pub variant_type: GhostProtocolVariantType,
+}
+
+impl GhostProtocolDiscriminant {
+    /// debug friendly identifier for the protocol variant type
+    pub fn id(&self) -> &str {
+        &self.id
+    }
+
+    /// whether this message variant type is directed to the actor or owner
+    pub fn destination(&self) -> &GhostProtocolDestination {
+        &self.destination
+    }
+
+    /// whether this message variant type is an event, request, or response
+    pub fn variant_type(&self) -> &GhostProtocolVariantType {
+        &self.variant_type
+    }
+}
+
+/// A GhostProtocol is most often an enum, the variants of which contain
+/// some additional meta-information indicating the message direction / type.
+pub trait GhostProtocol: 'static + Debug + Clone + Send + Sync {
+    /// a list of all variant types associated with this protocol
+    fn discriminant_list() -> &'static [GhostProtocolDiscriminant];
+
+    /// the meta information for this actual variant instance
+    fn discriminant(&self) -> &GhostProtocolDiscriminant;
+
+    // -- provided -- //
+
+    /// is this message destined for the actor?
+    fn is_to_actor(&self) -> bool {
+        match self.discriminant().destination() {
+            GhostProtocolDestination::Actor => true,
+            GhostProtocolDestination::Owner => false,
+        }
+    }
+
+    /// is this message destined for the owner?
+    fn is_to_owner(&self) -> bool {
+        !self.is_to_actor()
+    }
+
+    /// is this an event not expecting a response?
+    fn is_event(&self) -> bool {
+        if let GhostProtocolVariantType::Event = self.discriminant().variant_type() {
+            true
+        } else {
+            false
+        }
+    }
+
+    /// is this a request that expects a response?
+    fn is_request(&self) -> bool {
+        if let GhostProtocolVariantType::Request = self.discriminant().variant_type() {
+            true
+        } else {
+            false
+        }
+    }
+
+    /// is this a response to a previous request?
+    fn is_response(&self) -> bool {
+        if let GhostProtocolVariantType::Response = self.discriminant().variant_type() {
+            true
+        } else {
+            false
+        }
+    }
+}
+
+/// when a protocol handler received a callback for fulfilling a request
+/// it will follow this signature
+pub type GhostHandlerCb<'lt, T> = Box<dyn FnOnce(T) -> GhostResult<()> + 'lt + Send + Sync>;
```

Comment:
I'd renamed to `GhostHandlerCallback` as to not confuse with `GhostHandler`

Reply:
rust seems to like its shorthand... `fn` `impl` etc... I'd be cool changing this, but there's a bunch of other `...Cb` types in here, we should probably pick a pattern and stick with it.